File: lidar3d/host/scan3d/geometry.py
"""Umrechnung (Distanz, Scanwinkel, Gierwinkel) -> kartesische Punkte.

Aufbau: der 2D-LiDAR ist so montiert, dass seine Scanebene die Gierachse
enthaelt (Scanebene senkrecht). Ein Schrittmotor dreht ihn um diese Achse.
Damit ist eine Messung nichts anderes als eine Kugelkoordinate:

    alpha  = Polarwinkel, kommt vom LiDAR (0 deg = entlang +Z / nach oben)
    psi    = Azimut, kommt vom Schrittzaehler
    r      = Distanz

    x = r*sin(alpha)*cos(psi)
    y = r*sin(alpha)*sin(psi)
    z = r*cos(alpha)

Weil der LiDAR pro Umdrehung volle 360 Grad in seiner Ebene abdeckt, genuegt
ein Gierbereich von 180 Grad fuer die komplette Kugel. Deshalb braucht der
Aufbau keinen Schleifring.

In der Praxis sitzt das optische Zentrum nicht exakt auf der Gierachse. Die
beiden Versaetze (radial und axial) gehen vor der Drehung ein - siehe
docs/02-geometrie.md fuer die Kalibrierung.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Iterable, Iterator, List, Tuple

Point = Tuple[float, float, float]


@dataclass(frozen=True)
class MountGeometry:
    """Einbaulage des LiDAR relativ zur Gierachse. Alle Laengen in mm."""

    #: Abstand des optischen Zentrums von der Gierachse, senkrecht dazu.
    offset_radial_mm: float = 0.0
    #: Hoehe des optischen Zentrums entlang der Gierachse ueber dem Ursprung.
    offset_axial_mm: float = 0.0
    #: LiDAR-Winkel, der nach oben (+Z) zeigt.
    alpha_zero_deg: float = 0.0
    #: -1, wenn der LiDAR-Winkel entgegen der gewuenschten Richtung laeuft.
    alpha_sign: int = 1
    #: Gierwinkel, der als Azimut 0 gilt.
    yaw_zero_deg: float = 0.0
    #: -1, wenn der Schrittmotor entgegen der gewuenschten Richtung dreht.
    yaw_sign: int = 1

    def __post_init__(self) -> None:
        if self.alpha_sign not in (1, -1):
            raise ValueError("alpha_sign muss +1 oder -1 sein")
        if self.yaw_sign not in (1, -1):
            raise ValueError("yaw_sign muss +1 oder -1 sein")


@dataclass(frozen=True)
class RangeFilter:
    """Gueltigkeitsfenster fuer Messungen. Laengen in mm."""

    min_mm: float = 150.0  # unterhalb der Blindzone (Datenblatt: 50 mm) plus Reserve
    max_mm: float = 30000.0

    def accepts(self, distance_mm: float) -> bool:
        return distance_mm > 0 and self.min_mm <= distance_mm <= self.max_mm


def to_cartesian(
    distance_mm: float,
    alpha_deg: float,
    yaw_deg: float,
    mount: MountGeometry = MountGeometry(),
) -> Point:
    """Eine Messung in Meter-Weltkoordinaten (Z = Gierachse, nach oben)."""
    a = math.radians(mount.alpha_sign * (alpha_deg - mount.alpha_zero_deg))
    # Koordinaten in der Scanebene: u radial von der Achse weg, w entlang der Achse.
    u = mount.offset_radial_mm + distance_mm * math.sin(a)
    w = mount.offset_axial_mm + distance_mm * math.cos(a)

    psi = math.radians(mount.yaw_sign * (yaw_deg - mount.yaw_zero_deg))
    return (
        u * math.cos(psi) / 1000.0,
        u * math.sin(psi) / 1000.0,
        w / 1000.0,
    )


def project(
    samples: Iterable[Tuple[float, float, float]],
    mount: MountGeometry = MountGeometry(),
    rng: RangeFilter = RangeFilter(),
) -> Iterator[Point]:
    """(distance_mm, alpha_deg, yaw_deg)-Tripel in Punkte umrechnen und filtern."""
    for distance_mm, alpha_deg, yaw_deg in samples:
        if not rng.accepts(distance_mm):
            continue
        yield to_cartesian(distance_mm, alpha_deg, yaw_deg, mount)


def sweep_plan(
    yaw_span_deg: float,
    yaw_step_deg: float,
    scan_hz: float = 10.0,
    samples_per_second: float = 32000.0,
) -> dict:
    """Dauer, Punktzahl und Winkelaufloesung eines Sweeps vorausberechnen.

    ``yaw_step_deg`` ist der gewuenschte Abstand zwischen zwei Scanebenen.
    """
    if yaw_step_deg <= 0:
        raise ValueError("yaw_step_deg muss positiv sein")
    planes = yaw_span_deg / yaw_step_deg
    duration_s = planes / scan_hz
    samples_per_plane = samples_per_second / scan_hz
    return {
        "planes": planes,
        "duration_s": duration_s,
        "yaw_rate_deg_s": yaw_span_deg / duration_s if duration_s else 0.0,
        "samples_per_plane": samples_per_plane,
        "total_samples": planes * samples_per_plane,
        "in_plane_resolution_deg": 360.0 / samples_per_plane,
    }


def voxel_downsample(points: Iterable[Point], voxel_m: float) -> List[Point]:
    """Ein Punkt pro Voxel (erster Treffer gewinnt).

    Ohne das ist die Punktwolke an den Polen der Gierachse extrem dicht und am
    Aequator duenn - genau das erzeugt die typische Sanduhr-Form.
    """
    if voxel_m <= 0:
        raise ValueError("voxel_m muss positiv sein")
    seen = set()
    out: List[Point] = []
    for p in points:
        key = (
            math.floor(p[0] / voxel_m),
            math.floor(p[1] / voxel_m),
            math.floor(p[2] / voxel_m),
        )
        if key in seen:
            continue
        seen.add(key)
        out.append(p)
    return out

File: lidar3d/ios/LidarKit/Sources/LidarKit/Geometry.swift
// Kugelkoordinaten -> kartesische Punkte. Spiegelt host/scan3d/geometry.py.
//
// Der LiDAR liefert den Polarwinkel, der Schrittmotor den Azimut, also ist
// eine Messung direkt eine Kugelkoordinate. Herleitung und Kalibrierung
// stehen in docs/02-geometrie.md.

import Foundation
import simd

public struct MountGeometry: Sendable, Equatable {
    /// Abstand des optischen Zentrums von der Gierachse, senkrecht dazu, in mm.
    public var offsetRadialMm: Float
    /// Hoehe des optischen Zentrums entlang der Gierachse, in mm.
    public var offsetAxialMm: Float
    /// LiDAR-Winkel, der nach oben (+Z) zeigt.
    public var alphaZeroDeg: Float
    /// -1, wenn der LiDAR-Winkel entgegen der gewuenschten Richtung laeuft.
    public var alphaSign: Float
    /// Gierwinkel, der als Azimut 0 gilt.
    public var yawZeroDeg: Float
    public var yawSign: Float

    public init(offsetRadialMm: Float = 0,
                offsetAxialMm: Float = 0,
                alphaZeroDeg: Float = 0,
                alphaSign: Float = 1,
                yawZeroDeg: Float = 0,
                yawSign: Float = 1) {
        self.offsetRadialMm = offsetRadialMm
        self.offsetAxialMm = offsetAxialMm
        self.alphaZeroDeg = alphaZeroDeg
        self.alphaSign = alphaSign
        self.yawZeroDeg = yawZeroDeg
        self.yawSign = yawSign
    }

    public static let identity = MountGeometry()
}

public struct RangeFilter: Sendable, Equatable {
    public var minMm: Float
    public var maxMm: Float

    /// Untergrenze oberhalb der Blindzone (Datenblatt: 50 mm) plus Reserve.
    /// Obergrenze ist die Reichweite des C1 (12 m); der S2 kaeme auf 30 m.
    public init(minMm: Float = 150, maxMm: Float = 12000) {
        self.minMm = minMm
        self.maxMm = maxMm
    }

    @inline(__always)
    public func accepts(_ distanceMm: Float) -> Bool {
        distanceMm > 0 && distanceMm >= minMm && distanceMm <= maxMm
    }
}

private let degToRad = Float.pi / 180

/// Eine Messung in Meter-Weltkoordinaten. Z ist die Gierachse und zeigt nach oben.
@inline(__always)
public func toCartesian(distanceMm: Float,
                        alphaDeg: Float,
                        yawDeg: Float,
                        mount: MountGeometry = .identity) -> SIMD3<Float> {
    let a = mount.alphaSign * (alphaDeg - mount.alphaZeroDeg) * degToRad
    // Koordinaten in der Scanebene: u radial von der Achse weg, w entlang der Achse.
    // u darf negativ werden - genau deshalb genuegen 180 Grad Gieren.
    let u = mount.offsetRadialMm + distanceMm * sin(a)
    let w = mount.offsetAxialMm + distanceMm * cos(a)

    let psi = mount.yawSign * (yawDeg - mount.yawZeroDeg) * degToRad
    return SIMD3<Float>(u * cos(psi) / 1000, u * sin(psi) / 1000, w / 1000)
}

/// Eine Capsule in Punkte umrechnen und dabei ungueltige Messungen aussortieren.
public func project(capsule: CapsuleFrame,
                    mount: MountGeometry = .identity,
                    range: RangeFilter = RangeFilter(),
                    into points: inout [SIMD3<Float>]) {
    capsule.forEachSample { dist, alpha, yaw in
        guard range.accepts(dist) else { return }
        points.append(toCartesian(distanceMm: dist, alphaDeg: alpha, yawDeg: yaw, mount: mount))
    }
}

/// Dasselbe fuer den einfachen Scanmodus des C1.
public func project(scan: ScanFrame,
                    mount: MountGeometry = .identity,
                    range: RangeFilter = RangeFilter(),
                    into points: inout [SIMD3<Float>]) {
    scan.forEachSample { dist, alpha, yaw in
        guard range.accepts(dist) else { return }
        points.append(toCartesian(distanceMm: dist, alphaDeg: alpha, yawDeg: yaw, mount: mount))
    }
}

/// Dauer, Punktzahl und Aufloesung eines Sweeps vorausrechnen.
///
/// Der Scanner arbeitet Schritt fuer Schritt: je Ebene anfahren, einrasten,
/// genau eine LiDAR-Umdrehung erfassen. Die Dauer ist deshalb die Umdrehung
/// (1/scanHz) plus Zuschlag fuer Fahrt und Einrasten.
public struct SweepPlan: Sendable {
    public let planes: Float
    public let durationSeconds: Float
    public let secondsPerPlane: Float
    public let samplesPerPlane: Float
    public let totalSamples: Float
    public let inPlaneResolutionDeg: Float

    /// Vorbelegt mit den Werten des C1: 5000 Messungen/s bei 10 Hz, also rund
    /// 500 Punkte je Ebene und 0,72 Grad in der Ebene. Der S2 kaeme mit 32000
    /// auf 0,1125 Grad.
    public init(yawSpanDeg: Float = 180,
                yawStepDeg: Float = 1,
                scanHz: Float = 10,
                samplesPerSecond: Float = 5000,
                planeOverheadSeconds: Float = 0.05) {
        planes = yawSpanDeg / yawStepDeg
        secondsPerPlane = 1 / scanHz + planeOverheadSeconds
        durationSeconds = planes * secondsPerPlane
        samplesPerPlane = samplesPerSecond / scanHz
        totalSamples = planes * samplesPerPlane
        inPlaneResolutionDeg = 360 / samplesPerPlane
    }
}
